test/jdk/java/lang/instrument/RetransformWithMethodParametersTest.java

```diff
@@ -0,0 +1,243 @@
+ * @run compile -g -parameters RetransformWithMethodParametersTest.java
+ * @run shell MakeJAR.sh retransformAgent
+ *
+ * @run main/othervm -javaagent:retransformAgent.jar RetransformWithMethodParametersTest
+ */
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.lang.instrument.ClassFileTransformer;
+import java.lang.reflect.Method;
+import java.lang.reflect.Parameter;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.security.ProtectionDomain;
+import java.util.Arrays;
+
+import jdk.test.lib.JDKToolLauncher;
+import jdk.test.lib.process.ProcessTools;
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.util.ClassTransformer;
+
+/*
+ * The test verifies Instrumentation.retransformClasses() (and JVMTI function RetransformClasses)
+ * correctly handles MethodParameter attribute:
+ * 1) classfile bytes passed to transformers (and JVMTI ClassFileLoadHook event callback) contain the attribute;
+ * 2) the attribute is updated if new version has the attribute with different values;
+ * 3) the attribute is removed if new version doesn't contain the attribute.
+ */
+
+// See ClassTransformer.transform(int) comment for @1 tag explanations.
+class MethodParametersTarget {
+    // The class contains the only method, so we don't have issue with method sorting
+    // and ClassFileReconstituter should restore the same bytes as original classbytes.
+    public void method1(
+            int intParam1, String stringParam1 // @1 commentout
+            // @1 uncomment   int intParam2, String stringParam2
+            )
+    {
+        // @1 uncomment System.out.println(stringParam2);   // change CP
+    }
+}
+
+public class RetransformWithMethodParametersTest extends ATransformerManagementTestCase {
+
+    public static void main (String[] args) throws Throwable {
+        ATestCaseScaffold test = new RetransformWithMethodParametersTest();
+        test.runTest();
+    }
+
+    private String targetClassName = "MethodParametersTarget";
+    private String classFileName = targetClassName + ".class";
+    private String sourceFileName = "RetransformWithMethodParametersTest.java";
+    private Class targetClass;
+    private byte[] originalClassBytes;
+
+    private byte[] seenClassBytes;
+    private byte[] newClassBytes;
+
+    public RetransformWithMethodParametersTest() throws Throwable {
+        super("RetransformWithMethodParametersTest");
+
+        File origClassFile = new File(System.getProperty("test.classes", "."), classFileName);
+        log("Reading test class from " + origClassFile);
+        originalClassBytes = Files.readAllBytes(origClassFile.toPath());
+        log("Read " + originalClassBytes.length + " bytes.");
+    }
+
+    private void log(Object o) {
+        System.out.println(String.valueOf(o));
+    }
+
+    // Prints and verifies MethodParameters attribute using reflection.
+    private void verifyMethodParams(boolean expectedPresent, String... expectedNames) throws Throwable {
```

Comment:
Just a side comment, no intention to insist...
I think, a generalization with the `expectedPresent` parameter adds a little bit of complexity.
It feels that refactoring it into two methods - one for expected and another for non-expected would be more clear.

Reply:
Fixed